docs/internal/types.md

```diff
@@ -0,0 +1,359 @@
+
+## WARNING
+
+This is a preliminary document under heavy discussion.
+The ideas here are mostly right, the details, not necessarily.
+There will be a tutorial document that will be correct in all counts, but this isn't it.
+
+## Overview
+
+Verona is a statically typed language, with a structural and algebraic type system.
+
+The 'statically typed' bit means object types are always concrete and there is never implicit dynamic coercion like Perl or PHP.
+However, variable types can have a set of constraints instead of a concrete type, if and only if the possible types match the constraints.
+For example, variables can have abstract (interface) types if the objects bound to it are in a set of concrete (class) types that implement that interface.
+Another example is a type union, where a variable can be one of many types, but the code must deal with all possible types.
+The verification of these properties is done at compile time.
+
+The 'structural' bit means that open-world types can be defined by their shape: If an interface defines only a method `foo()` then any class that implements a method `foo()` implements this interface.
+This is in contrast to nominal typing, such as in Java, where an object implements an interface only if it adopts that interface.
+Objective-C supports both, with protocols defining nominal types but dynamic dispatch and explicit method checks providing structural typing.
+
+The 'algebraic' bit means that we can define types in terms of intersections or unions on other type expressions.
+For example, we can say `Integer | String` to define a set of constraints on a type that mean that it must either be an integer or a string (whatever they are: they may be concrete types, or they may be interfaces) or something like `Copyable & Hashable` to say any type that meets the constraints of the copyable and hashable interfaces.
+
+All types should be known at compile time and the compiler can assert that the objects and operations have the correct type at all times.
+Some types are inferred by the compiler, but there is no guessing: there's either one clear inferred type or it's an error.
+
+At a glance, Verona types are:
+ * Interfaces: abstract types that specify a set of fields and methods that an underlying concrete type must implement.
+   Interfaces may provide default implementations of methods that a class can adopt.
+ * Classes: concrete types.
+   A class may inherit from other classes but this does not make the class a subtype of the class from which it inherits.
+   Classes may explicitly adopt interfaces, in which case they adopt the default implementations from the interface, but conformance to an interface is entirely a property of the structural type.
+ * Capability rights: A Verona pointer is a *capability*: an unforgeable token of authority that permits access to an object. 
+   The set of rights that a capability holds are associated with a type, for example. `iso`, `mut`, `imm` and `readonly`.
```

Comment:
This bullet seems a little out of nowhere.  It might not hurt to make line 11 read "a structural and algebraic type system which, additionally, tracks access rights" and add an early gloss paragraph paralleling lines 19-24.  Something like...

```
Verona programs are, by construction, devoid of data races.
As part of ensuring that property, references to objects in Verona carry a 'capability' annotation, which conveys access rights to that object and global properties of the program.
For example, a reference may be marked as `iso`-lated: the reference grants full read and write access and assurance that no concurrent accesses are possible.
Or a reference may be `imm`-utable: only reads are possible and the system guarantees that all live references to this object are similarly read-only.
```

Reply:
I agree, please suggest that as a patch.